### DISCUSSION TITLE
"EFI stub: Exiting boot services and installing virtual address map..." need my files from ubuntu. HELP

### DISCUSSION THREAD

**Discussion by KingKilla94** [utmapp/UTM#3606] [UTM Help (macOS)]
I have already checked and according to people this is what fixes the problem 

When disk corruption happens, you might get a EFI stub: Exiting boot services and installing virtual address map... message on the next startup and the bootloader will hang. If your guest VM is based on Ubuntu, this can be solved by:

Turning off the VM
In the VM settings in the UTM management UI, under Display, switch the VM over to console-mode instead of full graphics.
Restart the VM and wait for it to boot into load into a BareBox bootloader.
Run fsck on the ubuntu disk partition (probably fsck /dev/mapper/ubuntu--vg-root if installed under default settings)
exit Barebox. This should try to restart the VM and load into the Ubuntu VM. If the Ubuntu VM loads, then your disk issues are fixed.
Turn off your VM again and switch back to full graphics in UTM.

for me it was /dev/mapper/ubuntu--vg-ubuntu--lv

however Whenever i : sudo fsck -y  /dev/mapper/ubuntu--vg-ubuntu--lv

it says its mounted, I try to unmount and it says "busy or something"

what can I do ?

**Comment by KingKilla94** [comment]
After more research I found out that apparently when you force quit , the distro gets corrupted. and now its so corrupted that I can not run it in full graphics mode, I can only run it on text only. 

anyone?

**Comment by adespoton** [comment]
You might want to change your QCOW flags on your drive images to stop this from happening... I'm not sure how you do that with the current UI though.

**Comment by KingKilla94** [comment]
Thanks bud , I have some files that I need. looks like im going to create a new virtual machine and just copy my files over.

**Comment by ismailyenigul** [comment]
Using UTM 3.1.5 on Macbook Pro M1 Pro chipset
installed Ubuntu 20.4 from ISO image.

I have two issues.
#1  too often lost IP connectivity to the VM
#2 getting disk corruption  while trying to fix issue #1 

I am getting disk corruption too often. time to time I lost IP connection to the VM.  in this case I stop the vm  and got "Do you want to force stop this VM and lose all unsaved data?" warning. There is no choose in most case. then vm file system is getting corrupted easily. I don't know it is QEMU problem or UTM.  I was not getting file system corruption on Virtualbox (on intel macbook)



<img width="699" alt="image" src="https://user-images.githubusercontent.com/2566232/158408538-d4b7f72e-af5b-4a58-94c6-c575aa18e6f5.png">

While bootloader in pending state, I could ssh to the box but it is too slows.
Here is the error messages

``` 
78.853704] EXT4-fs error (device dm-0): mb_free_blocks:1452: group 50, block 1657553:freeing already freed block (bit 19153); block bitmap corrupt.
[   78.853988] EXT4-fs error (device dm-0): ext4_mb_generate_buddy:744: group 50, block bitmap and bg descriptor inconsistent: 14468 vs 14469 free clusters

[ 1160.899799] EXT4-fs warning (device dm-0): ext4_dirblock_csum_verify:377: inode #527831: comm kubelet: No space for directory leaf checksum. Please run e2fsck -D.
[ 1160.899801] EXT4-fs error (device dm-0): __ext4_find_entry:1549: inode #527831: comm kubelet: checksumming directory block 0
[ 1229.890806] EXT4-fs error (device dm-0): ext4_validate_block_bitmap:385: comm containerd: bg 64: bad block bitmap checksum
[ 3517.568795] EXT4-fs warning (device dm-0): ext4_dirblock_csum_verify:377: inode #527831: comm kubelet: No space for directory leaf checksum. Please run e2fsck -D.
[ 3517.568798] EXT4-fs error (device dm-0): __ext4_find_entry:1549: inode #527831: comm kubelet: checksumming directory block 0
```


my disk and inode usage:

```
# df -h
Filesystem                         Size  Used Avail Use% Mounted on
udev                               926M     0  926M   0% /dev
tmpfs                              198M  1.8M  196M   1% /run
/dev/mapper/ubuntu--vg-ubuntu--lv   29G   11G   17G  40% /
tmpfs                              987M     0  987M   0% /dev/shm
tmpfs                              5.0M     0  5.0M   0% /run/lock
tmpfs                              987M     0  987M   0% /sys/fs/cgroup
/dev/vda2                          1.5G  111M  1.3G   9% /boot
/dev/loop1                          62M   62M     0 100% /snap/lxd/22530
/dev/vda1                          1.1G  5.1M  1.1G   1% /boot/efi
/dev/loop3                          38M   38M     0 100% /snap/snapd/14982
/dev/loop2                          61M   61M     0 100% /snap/lxd/21843
/dev/loop0                          58M   58M     0 100% /snap/core20/1332
/dev/loop4                          58M   58M     0 100% /snap/core20/1380
tmpfs                              198M     0  198M   0% /run/user/1000
tmpfs                              987M   12K  987M   1% /var/lib/kubelet/pods/4000521e-4f1b-4f8c-9c7f-2f821cc799c5/volumes/kubernetes.io~secret/calico-node-token-7m258
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/b2a58d8f696cfdb9503e49462c3c0d9d56b3118d9113eb1a5774a3eeef05d286/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/b2a58d8f696cfdb9503e49462c3c0d9d56b3118d9113eb1a5774a3eeef05d286/rootfs
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/384517ff952c9290a5ac97e77a295f3687a0b55bd0ca7be87e431b741c95a52d/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/384517ff952c9290a5ac97e77a295f3687a0b55bd0ca7be87e431b741c95a52d/rootfs
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/141194c27f9420fa42e519b2e2f9c6155beabd0e2e6d0ff23aa538fa9e3f4823/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/141194c27f9420fa42e519b2e2f9c6155beabd0e2e6d0ff23aa538fa9e3f4823/rootfs
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/db99da64a8a4bcb9cf9baa63507d36588f7143c9fbb59d774d5680f90ada9694/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/db99da64a8a4bcb9cf9baa63507d36588f7143c9fbb59d774d5680f90ada9694/rootfs
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/2fd80c7802038b712246ab11e13fb3d61d41b47dcac59a373ea48955589252ef/rootfs
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/808363648a673da21689c4190154addca7fc97637a043589b7e2db25cf4ddde0/rootfs
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/ac6b61a1a0ce490bbf0b6ac4655b9747968526ecf8a4224cc9a860e780c404a5/rootfs
/dev/loop5                          38M   38M     0 100% /snap/snapd/15183
tmpfs                              987M   12K  987M   1% /var/lib/kubelet/pods/7085576d-8de9-4c4c-825d-6be02f079939/volumes/kubernetes.io~projected/kube-api-access-qrtdw
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/78ff26f5ee4ad99d9bdb54a8a6f316ae747336265f6b21ac2bbd0731786dd888/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/78ff26f5ee4ad99d9bdb54a8a6f316ae747336265f6b21ac2bbd0731786dd888/rootfs
tmpfs                              987M   12K  987M   1% /var/lib/kubelet/pods/9df7c1ef-37d4-445c-b728-c12820e66021/volumes/kubernetes.io~projected/kube-api-access-nh94t
shm                                 64M     0   64M   0% /run/containerd/io.containerd.grpc.v1.cri/sandboxes/df1c729ae29de0e27afaa82e8a49109fc7114ca8b61964dd1317090cba62437b/shm
overlay                             29G   11G   17G  40% /run/containerd/io.containerd.runtime.v2.task/k8s.io/df1c729ae29de0e27afaa82e8a49109fc7114ca8b61964dd1317090cba62437b/rootfs

```

**Reply by Obligado1** [reply]
You got solution yet?

**Comment by ismailyenigul** [comment]
Not really. I tried a few times but did not work.  then switched to  https://multipass.run/

**Comment by iamredencio** [comment]
> I have already checked and according to people this is what fixes the problem
> 
> When disk corruption happens, you might get a EFI stub: Exiting boot services and installing virtual address map... message on the next startup and the bootloader will hang. If your guest VM is based on Ubuntu, this can be solved by:
> 
> Turning off the VM In the VM settings in the UTM management UI, under Display, switch the VM over to console-mode instead of full graphics. Restart the VM and wait for it to boot into load into a BareBox bootloader. Run fsck on the ubuntu disk partition (probably fsck /dev/mapper/ubuntu--vg-root if installed under default settings) exit Barebox. This should try to restart the VM and load into the Ubuntu VM. If the Ubuntu VM loads, then your disk issues are fixed. Turn off your VM again and switch back to full graphics in UTM.
> 
> for me it was /dev/mapper/ubuntu--vg-ubuntu--lv
> 
> however Whenever i : sudo fsck -y /dev/mapper/ubuntu--vg-ubuntu--lv
> 
> it says its mounted, I try to unmount and it says "busy or something"
> 
> what can I do ?

This actually worked for me! Thank you very much. I did not have a backup of this VM. Actually this was the backup. Was working in the original VM when both got corrupted. I blame the government.


The key was that you said to exit the Barebox. I literally typed in **exit** (Not just closing the VM) and when it asked me for my login I knew it was restored. Mine also was: fsck -y /dev/mapper/ubuntu--vg-ubuntu--lv
